Grammar.py

```diff
@@ -0,0 +1,96 @@
+from arpeggio import RegExMatch, Optional, ZeroOrMore, OneOrMore, EOF
+
+##############
+#Math Portion#
+##############
+
+#Positive or negative int
+def number():
+    return Optional("-"), RegExMatch('\d*\.\d*|\d+')
+
+#A variable/function identifier
+def var_ref():
+    return RegExMatch('[a-z][a-zA-Z_0-9]*')
+    
+#The base concept that can be used in an arithmetic_exression
+def factor():
+    return [number, fn_call, var_ref, ("(", arithmetic_expression, ")")]
+
+#One or more factors multiplied/divided together
+def mult_term():
+    return factor, ZeroOrMore(["*", "/"], factor)
+    
+#One or more terms added/subtracted together
+def arithmetic_expression():
+    return mult_term, ZeroOrMore(["+", "-"], mult_term)
+
+####################
+#Built in Functions#
+####################
+
+#Handles varaible assignment
+def var_decl():
+    return var_ref, "=", evaluatable
+
+#Handles variable declaration
+def var_let():
+    return "let", [var_decl, var_ref], ZeroOrMore(",", [var_decl, var_ref])
+
+#Handles boolean expressions ==, !=, >=, >, <=, <
+def boolean_expression():
+    return [(arithmetic_expression, ["==", "!=", ">=", ">", "<=", "<"], arithmetic_expression), "0", "1"]
+
+#Handles a print statement
+def print_func():
+    return "print", func_parameters
+
+#Handles code blocks statements
+def code_block():
+    return "{", ZeroOrMore(valid_line), "}"
+    
+#Handles if statements
+def if_statement():
+    return "if", boolean_expression, code_block, "else", code_block
+   
+###################
+#Cusotom Functions#
+###################
+   
+#Handles function assignment
+def fn_decl():
+    return [(Optional(var_ref, "="), ("fn", "(", Optional(var_ref), ZeroOrMore(",", var_ref), ")", code_block)), (var_ref, "=", fn_call)]
```

Comment:
This syntax has a couple of issues:

~~~ pr
          "fn", "(",
            Optional(var_ref),
            ZeroOrMore(",", var_ref),
          ")"
~~~

seems to allow `fn(,fred)` (note the leading comma).

# -1

More significant is the fact that this rule is also responsible for handling assignment of a function call to a variable.

The underlying problem here is that you're viewing the syntax as a set of statements, and not as a set of expressions that build into statements. If instead you had

~~~
fn_def <- "fn" params block
~~

then you could make `fn_def` an evaluable, and that then automatically allow it to be an rvalue.

The same would apply to assignments.

That's why the original grammar has:

~~~
statement = "let" variable_declaration
          | assignment
          | expr

expr  = "fn" function_definition
      | "if" if_expression
      | boolean_expression
      | arithmetic_expression

if_expression = expr brace_block ( "else" brace_block )?

assignment  = identifier "=" expr
~~~

It's a significant difference. In fact, looking at the original grammar, I now think that assignment should also be an expression.

Anyway, the code passes tests, but doesn't implemen the original grammar correctly.

# -5